File: src/main/java/net/nodus/database/common/MutableDocument.java
package net.nodus.database.common;

import java.time.Instant;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.LastModifiedDate;

@Getter
@Setter
public abstract class MutableDocument extends BaseDocument {

    @LastModifiedDate
    private Instant updatedAt = Instant.now();

    public void touch() {
        touch(Instant.now());
    }

    public void touch(Instant now) {
        this.updatedAt = now;
    }

    @Override
    public void softDelete(Instant now) {
        setDeletedAt(now);
        this.updatedAt = now;
    }

    @Override
    public void restore(Instant now) {
        setDeletedAt(null);
        this.updatedAt = now;
    }
}

File: src/main/kotlin/net/nodus/common/MutableDocument.kt
package net.nodus.common

import org.springframework.data.annotation.LastModifiedDate
import java.time.Instant

abstract class MutableDocument : BaseDocument() {
    @LastModifiedDate
    var updatedAt: Instant = Instant.now()

    fun touch(now: Instant = Instant.now()) {
        updatedAt = now
    }

    override fun softDelete(now: Instant) {
        deletedAt = now
        updatedAt = now
    }

    override fun restore(now: Instant) {
        deletedAt = null
        updatedAt = now
    }
}